src/3d/processing/qgsalgorithmtessellate.cpp

```diff
@@ -88,41 +89,76 @@ QgsFeatureList QgsTessellateAlgorithm::processFeature( const QgsFeature &feature
       f.clearGeometry();
     else
     {
-      const QgsRectangle bounds = f.geometry().boundingBox();
-      QgsTessellator t( bounds, false );
-
-      if ( f.geometry().isMultipart() )
+      const QgsGeometry inputGeometry = f.geometry();
+      bool tessellationComplete = false;
+#if (GEOS_VERSION_MAJOR==3 && GEOS_VERSION_MINOR>=11) || GEOS_VERSION_MAJOR>3
+      if ( !inputGeometry.constGet()->is3D() )
       {
-        const QgsMultiSurface *ms = qgsgeometry_cast< const QgsMultiSurface * >( f.geometry().constGet() );
-        for ( int i = 0; i < ms->numGeometries(); ++i )
+        // on supported GEOS versions we prefer to use GEOS GEOSConstrainedDelaunayTriangulation
+        // for 2D triangulation, as it's more stable and tolerant to situations like polygon
+        // holes touching an exterior ring vs the poly2tri based tessellation
+        const QgsGeometry triangulation = inputGeometry.constrainedDelaunayTriangulation();
+        if ( triangulation.isEmpty() && !inputGeometry.isEmpty() )
         {
-          const std::unique_ptr< QgsPolygon > p( qgsgeometry_cast< QgsPolygon * >( ms->geometryN( i )->segmentize() ) );
-          t.addPolygon( *p, 0 );
+          if ( !triangulation.lastError().isEmpty() )
+          {
+            feedback->reportError( QObject::tr( "Feature ID %1 could not be tessellated: %2" ).arg( f.id() ).arg( triangulation.lastError() ) );
```

Comment:
Out of curiosity, do you see scenarios when geos would fail but poly2tri would not? I'm wondering whether it'd be worth having a failed geos triangulation fallback to poly2tri here.

Reply:
> Out of curiosity, do you see scenarios when geos would fail but poly2tri would not

I don't. Poly2tri has been designed with much simpler geometries in mind vs typical spatial data. And geos is definitely designed and tested on those types of data! 🤣